core/services/relay/evm/functions/logpoller_wrapper.go

```diff
@@ -3,44 +3,87 @@ package functions
 import (
 	"context"
 	"sync"
+	"time"
 
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
 	"github.com/ethereum/go-ethereum/common"
 
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/client"
 	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/logpoller"
+	"github.com/smartcontractkit/chainlink/v2/core/gethwrappers/functions/generated/functions_coordinator"
+	"github.com/smartcontractkit/chainlink/v2/core/gethwrappers/generated/functions_router"
 	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/services/ocr2/plugins/functions/config"
 	evmRelayTypes "github.com/smartcontractkit/chainlink/v2/core/services/relay/evm/types"
+	"github.com/smartcontractkit/chainlink/v2/core/utils"
 )
 
 type logPollerWrapper struct {
-	routerContract  common.Address
-	contractVersion uint32
-	logPoller       logpoller.LogPoller
-	subscribers     map[string]evmRelayTypes.RouteUpdateSubscriber
-	mu              sync.RWMutex
-	lggr            logger.Logger
+	utils.StartStopOnce
+
+	routerContract      *functions_router.FunctionsRouter
+	pluginConfig        config.PluginConfig
+	client              client.Client
+	logPoller           logpoller.LogPoller
+	subscribers         map[string]evmRelayTypes.RouteUpdateSubscriber
+	activeCoordinator   common.Address
+	proposedCoordinator common.Address
+	nextBlock           int64
+	mu                  sync.Mutex
+	closeWait           sync.WaitGroup
+	stopCh              utils.StopChan
+	lggr                logger.Logger
 }
 
 var _ evmRelayTypes.LogPollerWrapper = &logPollerWrapper{}
 
-func NewLogPollerWrapper(routerContract common.Address, contractVersion uint32, logPoller logpoller.LogPoller, lggr logger.Logger) evmRelayTypes.LogPollerWrapper {
-	return &logPollerWrapper{
-		routerContract:  routerContract,
-		contractVersion: contractVersion,
-		logPoller:       logPoller,
-		subscribers:     make(map[string]evmRelayTypes.RouteUpdateSubscriber),
-		lggr:            lggr,
+func NewLogPollerWrapper(routerContractAddress common.Address, pluginConfig config.PluginConfig, client client.Client, logPoller logpoller.LogPoller, lggr logger.Logger) (evmRelayTypes.LogPollerWrapper, error) {
+	routerContract, err := functions_router.NewFunctionsRouter(routerContractAddress, client)
+	if err != nil {
+		return nil, err
 	}
+
+	return &logPollerWrapper{
```

Comment:
I see that we don't initalize the value for nextBlock.  Does that mean on the first call to LatestEvents, we are having the logPoller start polling logs from block zero?  Is that problematic?  When should be the starting block number from which the logPoller should start polling?

Reply:
Start() is initializing it synchronously, which happens before any LatestEvents() calls.